components/application-gateway/internal/proxy/proxy.go

```diff
@@ -138,17 +144,55 @@ func (p *proxy) addAuthorization(r *http.Request, cacheEntry *CacheEntry) apperr
 	return cacheEntry.CSRFTokenStrategy.AddCSRFToken(r)
 }
 
-func (p *proxy) addModifyResponseHandler(r *http.Request, id string, cacheEntry *CacheEntry) {
-	cacheEntry.Proxy.ModifyResponse = p.createModifyResponseFunction(id, r)
+func (p *proxy) addModifyResponseHandler(r *http.Request, id string, cacheEntry *CacheEntry) apperrors.AppError {
+	modifyResponseFunction, err := p.createModifyResponseFunction(id, r)
+	if err != nil {
+		return err
+	}
+
+	cacheEntry.Proxy.ModifyResponse = modifyResponseFunction
+	return nil
 }
 
-func (p *proxy) createModifyResponseFunction(id string, r *http.Request) func(*http.Response) error {
+func (p *proxy) createModifyResponseFunction(id string, r *http.Request) (func(*http.Response) error, apperrors.AppError) {
 	// Handle the case when credentials has been changed or OAuth token has expired
-	return func(response *http.Response) error {
-		retrier := newUnauthorizedResponseRetrier(id, r, p.proxyTimeout, p.createCacheEntry)
+	secondRequestBody, err := createSecondRequestBody(r)
+	if err != nil {
+		return nil, err
+	}
 
+	modifyResponseFunction := func(response *http.Response) error {
+		retrier := newUnauthorizedResponseRetrier(id, r, secondRequestBody, p.proxyTimeout, p.createCacheEntry)
 		return retrier.RetryIfFailedToAuthorize(response)
 	}
+
+	return modifyResponseFunction, nil
+}
+
+func createSecondRequestBody(r *http.Request) (io.ReadCloser, apperrors.AppError) {
```

Comment:
I would suggest the name `copyRequestBody`.

Reply:
Done!